### DISCUSSION TITLE
Adjust the judgment condition of nonlinear convergence

### DISCUSSION THREAD

**Discussion by AdelineHunter** [idaholab/moose#21539] [Q&A General] [answered]
Hello，

In some non-steady-state problems, some time steps may have less nonlinear residuals fall, while others may have more.  How should I give consideration to these two situations?

![image](https://user-images.githubusercontent.com/108536058/178015229-6d047a0d-629f-4535-9777-0a63d97660a2.png)

In other words, I want to make the `nl_rel_tol` that is difficult to converge less strict, and the `nl_rel_tol` that is easy to converge smaller. How can I set different `nl_rel_tol` at the same time

Thank you for your time.

**Comment by GiudGiud** [comment]
Hello

We currently dont support variable convergence criteria.
You can sort of hack it by setting min and max number of non linear iterations, but there's no first class support

The right way would be to rework the Executioner and the NLSystem classes to support functions for these convergence criteria

Guillaume

**Comment by lindsayad** [comment, marked as answer]
I think it might be more beneficial to figure out why the nonlinear solve is performing so poorly. Is your Jacobian inaccurate? You can test this with `-snes_test_jacobian`. Are you taking too large of timesteps such that your initial guess is too far from the root such that Newton's method performs poorly?

**Reply by GiudGiud** [reply]
also if you're debugging issues a particular timestep, it helps to create a checkpoint right before that time step (just add checkpoint=true to [Outputs]), then restart there to examine the issues. 

It could also be scaling issues. This page has a few potential issues listed
https://mooseframework.inl.gov/moose/application_usage/failed_solves.html

**Reply by AdelineHunter** [reply]
Thank you, Mr. Guillaume and Mr. lindsayad

Your advice is of great help to me.

I think maybe my problem is too complex, there are many problems like grid size, quality, and so on. Turbulence models may also be needed to help with convergence.

Thanks again
